src/contrib/mlir/compiler/op_lowerers.inc

```diff
@@ -20,26 +20,27 @@
 #endif
 
 #ifndef MLIR_LAST_OP
-#define MLIR_LAST_OP(OP) MLIR_OP(OP)
+#define MLIR_LAST_OP(OP, INPLACE) MLIR_OP(OP, INPLACE)
 #endif
 
-MLIR_OP(NGAddOp)
-MLIR_OP(NGArgMaxRedOp)
-MLIR_OP(NGArgMinRedOp)
-MLIR_OP(NGConcatOp)
-MLIR_OP(NGConvolutionOp)
-MLIR_OP(NGDivOp)
-MLIR_OP(NGDotOp)
-MLIR_OP(NGGatherOp)
-MLIR_OP(NGGreaterOp)
-MLIR_OP(NGLessOp)
-MLIR_OP(NGMulOp)
-MLIR_OP(NGMaxOp)
-MLIR_OP(NGMinOp)
-MLIR_OP(NGNegOp)
-MLIR_OP(NGReluOp)
-MLIR_OP(NGSubOp)
-MLIR_LAST_OP(NGReturnOp)
+        /* op name */         /* in-place safe ? */ 
+MLIR_OP(NGAddOp             , true                  )
+MLIR_OP(NGArgMaxRedOp       , false                 )
+MLIR_OP(NGArgMinRedOp       , false                 )
+MLIR_OP(NGConcatOp          , false                 )
+MLIR_OP(NGConvolutionOp     , false                 )
+MLIR_OP(NGDivOp             , true                  )
+MLIR_OP(NGDotOp             , false                 )
+MLIR_OP(NGGatherOp          , false                 )
+MLIR_OP(NGGreaterOp         , true                  )
+MLIR_OP(NGLessOp            , true                  )
+MLIR_OP(NGMulOp             , true                  )
+MLIR_OP(NGMaxOp             , true                  )
+MLIR_OP(NGMinOp             , true                  )
+MLIR_OP(NGNegOp             , true                  )
+MLIR_OP(NGReluOp            , true                  )
+MLIR_OP(NGSubOp             , true                  )
+MLIR_LAST_OP(NGReturnOp     , false                 )
```

Comment:
I was wondering if we could model this as a MLIR attribute of the op. Something like `IsImplaceSafe` that we add to tablegen. Would it make sense?

Reply:
Tablegen doesn't allow declaring C++ (virtual) inheritance unfortunately. We would still have to do `isa<>() + `cast<>()` followed by `isInPlaceSafe`, which would be redundant at this point.

Reply:
I'm not sure I follow but, ok, we can address that later

Reply:
If I add an `IsInPlaceSafe` helper to table-gen defs, it will be a replicated method across each C++ class. It won't be inherited from a base class. So eventually given an Op*, I still have to figure out its dynamic type to be able to call the helper (no virtual base implementation exists).

Reply:
I think that should be fine. Code replication when using tablegen is expected and it should be a problem since it's auto-generated. Filtering ops using TTI is also fine. My comment was more about consistently modeling things using MLIR features (i.e., tablegen and attributes, in this case) since things can get complicated quickly for an `.inc` file